File: First Missing Positive.java
public class Solution {
    public int firstMissingPositive(int[] A) {
        for (int i = 0; i < A.length; i++) {
            while (A[i] >= 1 && A[i] <= A.length) {
                // swap A[A[i] - 1] with A[i]
                int temp = A[i];
                if (temp == A[temp-1])
                    break;
                A[i] = A[temp - 1];
                A[temp - 1] = temp;
            }
        }
        for (int i = 0; i < A.length; i++) {
            if (A[i] != i+1)
                return i+1;
        }
        return A.length + 1;
    }
}

File: First Missing Positive.py
class Solution:
	# @param A, a list of integers
	# @return an integer
	def firstMissingPositive(self, A):
		for i in xrange(len(A)):
			while (A[i] - 1 >= 0 and A[i] - 1 < len(A)):
				n = A[i]
				# swap A[i] and A[n-1]
				if (n == A[n-1]):
					break;
				tmp = A[i]
				A[i] = A[n-1]
				A[n-1] = tmp
		for i in xrange(len(A)):
			if (A[i] != i+1):
				return i+1
		return len(A)+1